hashintel/hash
# H-2595: Enable browser plugin to provide website content requested by Flow step

## 🌟 What is the purpose of this PR?

This PR introduces the ability for Flow steps to request external input when they encounter a need for it, e.g. clarification or further advice from a human, or the content of a webpage that requires authentication from the user's browser (via the browser plugin).

It also implements the resolution of requests for website content, via the communication between the browser plugin and the websocket in the Node API.

The process by which these requests and responses occur is:
  1. Activity sends a Signal to the Workflow requesting data from the outside world when it encounters a need for it
  2. The outside world is polling for these requests (resolved by finding the signals in the event history, for now) – in the case of 'get website content', the Node API is polling for run data, which now has requests added to it
  3. The request is resolved – in this case, the browser plugin picks up the request via the websocket, opens a window, loads the page, and sends the content back as a websocket message
  4. The outside world sends a Signal back to the Workflow with the requested data – in this case, the Node API sends the Signal once it receives a message from the browser plugin
  5. The Workflow allows the response to be accessed via a Query, which the activity is polling for
 
The PR also:
1. Handles in the UI the case where a workflow has been cancelled (i.e. show it as cancelled / red colours etc)
2. Consolidates the existing 'get webpage' functions into one, and makes it more resilient to errors and scraping detection
3. Adds some more testing entity types to the test script (`npx tsx apps/hash-api/src/seed-data/seed-flow-test-types.ts`)
4. Adds another example flow definition, which researches some entities and then writes them to a Google Sheet (i.e. the existing research task without needing to have some question at the end to feed into the question analysis step)

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## 🐾 Next steps

- H-2572 tracks implementation of requests for human input


## 🛡 What tests cover this?



- None yet

## ❓ How to test this?



1. Run `yarn dev` in the browser plugin and make sure it's in your browser and you're logged in
2. Start a research job with a brief that will cause a visit to LinkedIn, e.g. 'look at LinkedIn for info about Bill Gates'
